app/utils/SQL.scala

```diff
@@ -0,0 +1,251 @@
+    SqlToken(sql = s"(${sqlValues.map(_.getPlaceholder()).mkString(", ")})", values = sqlValues)
+  }
+
+  def tupleList(values: List[List[Any]]): SqlToken = {
+    val sqlValueLists = values.map(list => list.map(SqlValue.makeSqlValue))
+    SqlToken(sql = sqlValueLists.map(list => s"(${list.map(_.getPlaceholder()).mkString(", ")})").mkString(", "),
+             values = sqlValueLists.flatten)
+  }
+
+  def raw(s: String) = SqlToken(s)
+
+  def empty(): SqlToken = raw("")
+
+  def identifier(id: String): SqlToken = raw('"' + id + '"')
+}
+
+trait SqlValue {
+  def setParameter(pp: PositionedParameters): Unit
+
+  def getPlaceholder(): String = "?"
+
+  def debug(): String
+}
+
+object SqlValue {
+
+  def makeSqlValue(p: Any): SqlValue =
+    p match {
+      case x: SqlValue => x
+      case x: String   => StringValue(x)
+      case x: Option[_] =>
+        x match {
+          case Some(y) => makeSqlValue(y)
+          case None    => NoneValue()
+        }
+      case x: Short    => ShortValue(x)
+      case x: Int      => IntValue(x)
+      case x: Long     => LongValue(x)
+      case x: Float    => FloatValue(x)
+      case x: Double   => DoubleValue(x)
+      case x: Boolean  => BooleanValue(x)
+      case x: Instant  => InstantValue(x)
+      case x: Duration => DurationValue(x)
+      case x: ObjectId => ObjectIdValue(x)
+      case x: JsValue  => JsonValue(x)
+    }
+}
+
+case class StringValue(v: String) extends SqlValue {
+  override def setParameter(pp: PositionedParameters): Unit = pp.setString(v)
+
+  override def debug(): String = "'" + v + "'"
+}
+
+case class ShortValue(v: Short) extends SqlValue {
+  override def setParameter(pp: PositionedParameters): Unit = pp.setShort(v)
+
+  override def debug(): String = s"$v"
+}
+
+case class IntValue(v: Int) extends SqlValue {
+  override def setParameter(pp: PositionedParameters): Unit = pp.setInt(v)
+
+  override def debug(): String = s"$v"
+}
+
+case class LongValue(v: Long) extends SqlValue {
+  override def setParameter(pp: PositionedParameters): Unit = pp.setLong(v)
+
+  override def debug(): String = s"$v"
+}
+
+case class FloatValue(v: Float) extends SqlValue {
+  override def setParameter(pp: PositionedParameters): Unit = pp.setFloat(v)
+
+  override def debug(): String = s"$v"
+}
+
+case class DoubleValue(v: Double) extends SqlValue {
+  override def setParameter(pp: PositionedParameters): Unit = pp.setDouble(v)
+
+  override def debug(): String = s"$v"
+}
+
+case class BooleanValue(v: Boolean) extends SqlValue {
+  override def setParameter(pp: PositionedParameters): Unit = pp.setBoolean(v)
+
+  override def debug(): String = s"$v"
+}
+
+case class InstantValue(v: Instant) extends SqlValue {
+  override def setParameter(pp: PositionedParameters): Unit = pp.setTimestamp(v.toSql)
+
+  override def getPlaceholder(): String = "?::TIMESTAMPTZ"
+
+  override def debug(): String = s"'${v.toString}'::TIMESTAMPTZ"
+}
+
+case class DurationValue(v: Duration) extends SqlValue {
```

Comment:
I think we should use FiniteDuration – the generic Duration can also hold infinity, which will lead to unexpected behavior when calling length. I don’t think we currently have any semantics where infinity is needed in webKnossos.